### PR TITLE
Add network policy examples

### PR DESCRIPTION
**Pull request by rinckm**
Add network policy examples

### Dependency release notes

N/A

### Submitter checklist

- [ ] N/A Change has been tested (on a back-end cluster)
- [x] [changelog.yaml] with upgrade notes are prepared and appropriate for the audience affected by the change (users or developer, depending on the change).
- [x] Semantic version diffed against [last release][releases] and updated accordingly. In this project the version has to be maintained here:
    - [/charts/steward/Chart.yaml](https://github.com/SAP/stewardci-core/blob/master/charts/steward/Chart.yaml) (`version` and `appVersion`)

In case dependencies have been updated:

N/A

### Reviewer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There is at least one approval for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] The Pull Request title is understandable and reflects the changes well
- [x] The Pull Request description is understandable and well documented
- [x] 'Upgrade notes' are documented in changelog.yaml (if required)
- [x] [changelog.yaml] entry for this Pull Request has been added
    - [x] Changelog entry contains all required information

[changelog.yaml]: https://github.com/SAP/stewardci-core/changelog.yaml
[releases]: https://github.com/SAP/stewardci-core/releases

### CODE REVIEW

**Comment by alxsap** [comment]
Will those examples stay stable or will we need to keep them in sync with the productive network policies (structure)?

**Comment by anbrsap** [comment]
> Will those examples stay stable or will we need to keep them in sync with the productive network policies (structure)?

The examples are generic. They don't have to correspond to particular network policies or network profiles used by specific Steward deployments of SAP.

File: docs/examples/network_policies/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Network policy examples
+
+## Valid policy configuration
+
+To run this examples apply the `policies.yaml` to your setward namespace.
```
Comment (anbrsap) [inline docs/examples/network_policies/README.md:5]:
```suggestion
To run these examples apply `policies.yaml` to your Steward system namespace, e.g. `steward-system`:
```

**Comment by anbrsap** [comment]
I would rather use the term "network profile" instead of "network policy", because this is what you configure in the pipeline run.

Currently a network profile determines only the network policy to be used, but in the future other network-related configuration might be covered by it.

File: docs/examples/network_policies/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+The result is a run failing with a content error.
+
+```
+NAME                             STARTED   FINISHED   STATUS     RESULT          MESSAGE
+network-profile-blocked-qgjfp    **        **         finished   error_content   Command ['git' 'clone' 'https://github.com/SAP-samples/stewardci-example-pipelines' '.'] failed w...
+```
+
+### Network policy with unknown policy
+
+This exaple pipeline run tries to use a non existing network policy.
```
Comment (anbrsap) [inline docs/examples/network_policies/README.md:42]:
```suggestion
This example pipeline run tries to use a nonexistent network policy.

```

File: docs/examples/network_policies/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+```
+kubectl -n steward-system apply -f pipelinerun_network_unknown.yaml
+```
+
+The result is a run failing with a config error.
```
Comment (anbrsap) [inline docs/examples/network_policies/README.md:48]:
```suggestion
The pipeline run result is `error_config`, which indicates that the pipeline run specification was erroneous.
```

File: docs/examples/network_policies/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+```
+NAME                             STARTED   FINISHED   STATUS     RESULT          MESSAGE
+network-profile-unknown-9p8v8    **        **         finished   error_config    ERROR: preparing failed ...
+```
+
+## Invalid policy configuration
+
+If the definition of the network policy config map is inconsistent (e.g. not existing default policy), you will get an infrastructure error.
```
Comment (anbrsap) [inline docs/examples/network_policies/README.md:57]:
```suggestion
If the definition of the network policy config map is inconsistent (e.g. not existing default policy), all pipeline runs will fail with result `error_infra`.

```

File: docs/examples/network_policies/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+To run this example apply the `policies_inconsistent.yaml` to your setward namespace.
```
Comment (anbrsap) [inline docs/examples/network_policies/README.md:59]:
```suggestion
To run this example apply `policies_inconsistent.yaml` to your Steward system namespace, e.g.:
```

File: docs/examples/network_policies/README.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+```
+kubectl -n steward-system apply -f policies.yaml
+```
+
+### Network policy with internet access
+
+This example pipeline run uses a network policy which allows internet access.
+
+```
+kubectl -n steward-system apply -f pipelinerun_network_internet.yaml
```
Comment (anbrsap) [inline docs/examples/network_policies/README.md:16]: The pipeline run should not be created in the Steward system namespace. Although it technically works, it gives a wrong impression to the reader of this page.

**Review by anbrsap — CHANGES_REQUESTED** [review]

File: docs/examples/network_policies/README.md (hunk 6 above)
Comment (rinckm) [reply docs/examples/network_policies/README.md]: See https://github.com/SAP/stewardci-core/pull/188/commits/8a4d0a17c8336f18f8b65d8a44f74fe927b9ed8d

**Comment by alxsap** [comment]
#190 will fix the voter. Merge master to this PR, once #190 is merged.

File: changelog.yaml [hunk 7]
```diff
@@ -51,6 +51,10 @@
 - version: NEXT
   date: TBD
   changes:
+  - type: doucmentation
+    impact: patch
+    title: Add network policy examples
+    description: Add examples for pipeline runs with configured network profiles.
```
Comment (rinckm) [inline changelog.yaml:57]:
```suggestion
    description: Add examples for pipeline runs with configured network profiles.
    pullRequestNumber: 188
```

File: changelog.yaml [hunk 8]
```diff
@@ -51,6 +51,11 @@
 - version: NEXT
   date: TBD
   changes:
+  - type: doucmentation
```
Comment (rinckm) [inline changelog.yaml:54]:
```suggestion
  - type: documentation
```

File: docs/examples/network_profile/README.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
+# Network profile examples
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:1]: Use title case

File: docs/examples/network_profile/README.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
+# Network profile examples
+
+To run this examples you need to [create a Tenant](../README.md#tenant) and [set the environment variable](../README.md#pipelinerun) `TENANT_NAMESPACE` accordingly.
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:3]:
```suggestion
To run these examples you need to [create a Tenant](../README.md#tenant) and [set the environment variable](../README.md#pipelinerun) `TENANT_NAMESPACE` accordingly.

```

File: docs/examples/network_profile/profiles_inconsistent.yaml [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+              app.kubernetes.io/instance: malware-scan-proxy
+      ingress:
+      - # rule: allow access from all pods in the same namespace
+        from:
+        - podSelector: {}
+kind: ConfigMap
+metadata:
+  labels:
+  name: steward-pipelineruns-network-policies
+  namespace: steward-system
```
Comment (anbrsap) [inline docs/examples/network_profile/profiles_inconsistent.yaml:98]: Move these entries above the `data` section for better understandability.

File: docs/examples/network_profile/profiles.yaml [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+              app.kubernetes.io/instance: malware-scan-proxy
+      ingress:
+      - # rule: allow access from all pods in the same namespace
+        from:
+        - podSelector: {}
+kind: ConfigMap
+metadata:
+  labels:
+  name: steward-pipelineruns-network-policies
+  namespace: steward-system
```
Comment (anbrsap) [inline docs/examples/network_profile/profiles.yaml:98]: Move these entries above the `data` section for better understandability.

Comment (anbrsap) [inline docs/examples/network_profile/profiles.yaml:98]: The Steward system namespace might be different, so we should not include it here. Instead, it should be specified via kubectl argument.

File: docs/examples/network_profile/profiles_inconsistent.yaml (hunk 11 above)
Comment (anbrsap) [inline docs/examples/network_profile/profiles_inconsistent.yaml:98]: ditto

File: docs/examples/network_profile/README.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+kubectl -n  $TENANT_NAMESPACE apply -f pipelinerun_network_internet.yaml
+```
+
+The result is a successful run.
+```
+NAME                             STARTED   FINISHED   STATUS     RESULT          MESSAGE
+network-profile-internet-qgqvf   **        **         finished   success         Pipeline completed with result: SUCCESS
+```
+
+### Network profile with no network access
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:27]:
```suggestion
### Network profile without network access

```

File: docs/examples/network_profile/README.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+This example pipeline run uses a network profile which allows no network access.
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:29]:
```suggestion
This example pipeline run uses a network profile which blocks any network communication.

```

File: docs/examples/network_profile/README.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+The result is a run failing with a content error.
+
+```
+NAME                             STARTED   FINISHED   STATUS     RESULT          MESSAGE
+network-profile-blocked-qgjfp    **        **         finished   error_content   Command ['git' 'clone' 'https://github.com/SAP-samples/stewardci-example-pipelines' '.'] failed w...
+```
+
+### Network profile with unknown profile
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:42]:
```suggestion
### Unknown network profile

```

File: docs/examples/network_profile/README.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+If the definition of the network profile config map is inconsistent (e.g. not existing default profile), all pipeline runs will fail with result `error_infra`.
+
+
+To run this example apply `profiles.yaml_inconsistent.yaml` to your Steward system namespace, e.g.:
+
+```
+kubectl -n steward-system apply -f profiles.yaml_inconsistent.yaml
+kubectl -n  $TENANT_NAMESPACE apply -f pipelinerun_network_internet.yaml
+```
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:68]:
````suggestion
To run this example apply `profiles_inconsistent.yaml` to your Steward system namespace, e.g.:

```
kubectl -n steward-system apply -f profiles_inconsistent.yaml
kubectl -n  "$TENANT_NAMESPACE" apply -f pipelinerun_network_internet.yaml
```

````

File: docs/examples/network_profile/README.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+The pipeline run result is `error_config`, which indicates that the pipeline run specification was erroneous.
+
+```
+NAME                             STARTED   FINISHED   STATUS     RESULT          MESSAGE
+network-profile-unknown-9p8v8    **        **         finished   error_config    ERROR: preparing failed ...
+```
+
+## Invalid profile configuration
+
+If the definition of the network profile config map is inconsistent (e.g. not existing default profile), all pipeline runs will fail with result `error_infra`.
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:60]: We should mention the effect of feature flag `RetryOnInvalidPipelineRunsConfig`.

File: docs/examples/network_profile/profiles_inconsistent.yaml [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+            # attached. Therefore we cannot select it here.
+            # If we can set labels on the namespace, make this selector more
+            # specific.
+          podSelector:
+            matchLabels:
+              app.kubernetes.io/instance: malware-scan-proxy
+      ingress:
+      - # rule: allow access from all pods in the same namespace
+        from:
+        - podSelector: {}
```
Comment (anbrsap) [inline docs/examples/network_profile/profiles_inconsistent.yaml:93]: Do we need such long data section just for creating an erroneous config? How should the reader spot the error?

**Review by anbrsap — CHANGES_REQUESTED** [review]

File: docs/examples/network_profile/profiles_inconsistent.yaml (hunk 18 above)
Comment (rinckm) [reply docs/examples/network_profile/profiles_inconsistent.yaml:93]: diff profiles.yaml profiles_inconsistent.yaml

Comment (rinckm) [reply docs/examples/network_profile/profiles_inconsistent.yaml:93]: Did the diff as small as possible by intention.

File: docs/examples/network_profile/README.md (hunk 17 above)
Comment (rinckm) [reply docs/examples/network_profile/README.md]: Done

File: docs/examples/network_profile/profiles.yaml (hunk 12 above)
Comment (rinckm) [reply docs/examples/network_profile/profiles.yaml]: Removed

Comment (rinckm) [reply docs/examples/network_profile/profiles.yaml]: Done

File: docs/examples/network_profile/profiles_inconsistent.yaml (hunk 11 above)
Comment (rinckm) [reply docs/examples/network_profile/profiles_inconsistent.yaml]: Done

File: docs/examples/network_profile/README.md (hunk 9 above)
Comment (rinckm) [reply docs/examples/network_profile/README.md]: Done

File: docs/examples/network_profile/profiles.yaml [hunk 19]
```diff
@@ -0,0 +1,97 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
```
Comment (anbrsap) [inline docs/examples/network_profile/profiles.yaml:4]:
```suggestion
```

File: docs/examples/network_profile/profiles.yaml [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+      - # rule: allow access to malware scan proxy
+        to:
+        - namespaceSelector: {}
+            # The namespace is 'malware-scan-proxy', but has no labels
+            # attached. Therefore we cannot select it here.
+            # If we can set labels on the namespace, make this selector more
+            # specific.
+          podSelector:
+            matchLabels:
+              app.kubernetes.io/instance: malware-scan-proxy
```
Comment (anbrsap) [inline docs/examples/network_profile/profiles.yaml:93]: Please remove that!

File: docs/examples/network_profile/README.md [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+```
+kubectl -n steward-system apply -f profiles.yaml
+```
+
+### Network profile with internet access
+
+This example pipeline run uses a network profile which allows internet access.
+
+```
+kubectl -n  $TENANT_NAMESPACE apply -f pipelinerun_network_internet.yaml
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:19]:
```suggestion
kubectl -n "$TENANT_NAMESPACE" apply -f pipelinerun_network_internet.yaml

```

File: docs/examples/network_profile/README.md [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+### Network profile without network access
+
+
+This example pipeline run uses a network profile which blocks any network communication.
+
+
+```
+kubectl -n  $TENANT_NAMESPACE apply -f pipelinerun_network_blocked.yaml
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:35]:
```suggestion
kubectl -n "$TENANT_NAMESPACE" apply -f pipelinerun_network_blocked.yaml

```

File: docs/examples/network_profile/README.md [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+## Invalid profile configuration
+
+If the definition of the network profile config map is inconsistent (e.g. not existing default profile), all pipeline runs will fail with result `error_infra`. This behaviour could be changed by the [feature flag RetryOnInvalidPipelineRunsConfig](../../../charts/steward/README.md#list-of-defined-feature-flags).
+
+
+To run this example apply `profiles_inconsistent.yaml` to your Steward system namespace, e.g.:
+
+```
+kubectl -n steward-system apply -f profiles_inconsistent.yaml
+kubectl -n  "$TENANT_NAMESPACE" apply -f pipelinerun_network_internet.yaml
```
Comment (anbrsap) [inline docs/examples/network_profile/README.md:71]:
```suggestion
kubectl -n "$TENANT_NAMESPACE" apply -f pipelinerun_network_internet.yaml

```

File: docs/examples/network_profile/profiles.yaml (hunk 19 above)
Comment (rinckm) [reply docs/examples/network_profile/profiles.yaml]: removed

File: docs/examples/network_profile/profiles.yaml (hunk 20 above)
Comment (rinckm) [reply docs/examples/network_profile/profiles.yaml:92]: removed

File: docs/examples/network_profile/README.md (hunk 21 above)
Comment (rinckm) [reply docs/examples/network_profile/README.md]: fixed

File: docs/examples/network_profile/README.md (hunk 22 above)
Comment (rinckm) [reply docs/examples/network_profile/README.md]: fixed

File: docs/examples/network_profile/README.md (hunk 23 above)
Comment (rinckm) [reply docs/examples/network_profile/README.md]: fixed

File: docs/examples/network_profile/profiles.yaml (hunk 20 above)
Comment (anbrsap) [reply docs/examples/network_profile/profiles.yaml]: No, it was still there. I remove it.

**Review by anbrsap — APPROVED** [review]

**Comment by anbrsap** [comment]
**Squash before merge!**

**Pull request merged**